### PR TITLE
Feature/tech 779

### PR DESCRIPTION
**Pull request by lruzicki**
Ticket:
https://govstack-global.atlassian.net/browse/TECH-779

### CODE REVIEW

File: govstack_api/building_blocks/bb_digital_registries/registries/insuree_registry.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+from django.apps import apps
+from govstack_api.building_blocks.bb_digital_registries.registries.base_registry import BaseRegistry
+from govstack_api.graphql_api_client import GrapheneClient
+from insuree.schema import Query, Mutation
+
+
+config = apps.get_app_config('govstack_api')
+
+
+def get_insurees_query(arguments_with_variables: str = "", fetched_fields: str = "") -> str:
```
Comment (lruzicki) [inline govstack_api/building_blocks/bb_digital_registries/registries/insuree_registry.py]: Do we want to move it to other folder to make it more readable?

File: govstack_api/building_blocks/bb_digital_registries/registries/insuree_registry.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                else:
+                    json_ext = {}
+
+        for special_field in ['BirthCertificateID', 'PersonalData']:
+            if special_field in json_ext:
+                mapped_data[special_field] = json_ext[special_field]
+
+        return mapped_data
+
+    def extract_records(self, result, only_first=True):
```
Comment (lruzicki) [inline govstack_api/building_blocks/bb_digital_registries/registries/insuree_registry.py]: those and methods below will be moved to helpers.py since they can be reusable with other models

File: govstack_api/apps.py [hunk 3]
```diff
@@ -4,3 +4,7 @@
 class TestHarnessApiConfig(AppConfig):
     default_auto_field = 'django.db.models.BigAutoField'
     name = 'govstack_api'
+
+    IM_CLIENT = 'eGovStack/GOV/90000009/digitalregistries'
```
Comment (dborowiecki) [inline govstack_api/apps.py]: I think this could be loaded from .env instead of being hardcoded.

File: govstack_api/building_blocks/bb_digital_registries/registries/base_registry.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+        pass
+
+    def update_record(self):
+        pass
+
+    def delete_record(self):
+        pass
+
+    def check_if_record_exists(self):
+        pass
```
Comment (dborowiecki) [inline govstack_api/building_blocks/bb_digital_registries/registries/base_registry.py]: I think you can use typehint `Protocol` inheritance in here to create implicit interface (it will be more pythinic than using abstract classes). I wouldn't call this one 'BaseRegistry' as there is no base funcionality provided.

File: govstack_api/graphql_api_client.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class GrapheneClient:
+    def __init__(self, request, query, mutation):
+        self._client = Client(Schema(query=query, mutation=mutation))
+        self.context = self.get_context(request)
+
+    def create_base_context(self):
+        user = mock.Mock(is_anonymous=False)
+        user.has_perm = mock.MagicMock(return_value=False)
```
Comment (dborowiecki) [inline govstack_api/graphql_api_client.py]: why is this part needed? User is always replaced with request.user, I think it an be removed

File: govstack_api/middleware.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        im_client = config.IM_CLIENT
+        im_header = request.META.get('HTTP_INFORMATION_MEDIATOR_CLIENT')
+        if im_header != im_client:
+            return HttpResponse('Unauthorized', status=401)
+        else:
+            # self.authenticate_user(request)
+            pass
+
+        return view_func(request, *args, **kwargs)
+    return _wrapped_view
```
Comment (dborowiecki) [inline govstack_api/middleware.py:28]: Is this function needed when we InformationmediatorMiddleware is in use?

File: govstack_api/middleware.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        if request.headers.get('Information-Mediator-Client', None):
+            config = apps.get_app_config('govstack_api')
+            im_client = config.IM_CLIENT
+            im_header = request.META.get('HTTP_INFORMATION_MEDIATOR_CLIENT')
+            if im_header != im_client:
+                return HttpResponse('Unauthorized', status=401)
```
Comment (dborowiecki) [inline govstack_api/middleware.py]: We could add a note that this comparison is only temporary and awaiting  information mediator setup.

File: govstack_api/middleware.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                self.authenticate_user(request)
+        response = self.get_response(request)
+        return response
+
+    def get_client(self, schema, query, mutation):
+        return Client(schema=schema(query=query, mutation=mutation))
+
+    def create_base_context(self):
+        user = mock.Mock(is_anonymous=False)
+        user.has_perm = mock.MagicMock(return_value=False)
```
Comment (dborowiecki) [inline govstack_api/middleware.py]: Same as for gql client, this user is always replaced and it can be removed

File: govstack_api/building_blocks/bb_digital_registries/registries/insuree_registry.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+            }}
+          ) {{
+            clientMutationId
+            internalId
+          }}
+        }}
+        '''
+
+
+class InsureeRegistry(BaseRegistry):
```
Comment (dborowiecki) [inline govstack_api/building_blocks/bb_digital_registries/registries/insuree_registry.py]: While I like new architecture for this module, the InsureeRegistry still has multiple hardcoded parts like fields mapping and gql query templates. Finally we need to have possibility to define registries based on the minimal setup and ideally no coding, only configuration.

**Review by dborowiecki — COMMENTED** [review]
Left some comments

File: govstack_api/middleware.py (hunk 6 above)
Comment (lruzicki) [reply govstack_api/middleware.py:28]: We have decorator function and middleware since in testing there is issue with accessing swagger endpoint that is used to for endpoint testing.

File: govstack_api/apps.py (hunk 3 above)
Comment (lruzicki) [reply govstack_api/apps.py]: done

File: govstack_api/graphql_api_client.py (hunk 5 above)
Comment (lruzicki) [reply govstack_api/graphql_api_client.py:14]: You are right. I will delete "def create_base_context()" and replace it with `context = SimpleNamespace(user=request.user)`

File: govstack_api/middleware.py (hunk 7 above)
Comment (lruzicki) [reply govstack_api/middleware.py]: done

File: govstack_api/graphql_api_client.py (hunk 5 above)
Comment (lruzicki) [reply govstack_api/graphql_api_client.py]: I think we can simplify it to:
```
    def get_context(self, request):
        return SimpleNamespace(user=request.user)
```

File: govstack_api/middleware.py (hunk 8 above)
Comment (lruzicki) [reply govstack_api/middleware.py]: Yeah I will change it to:
```
    def get_context(self, request):
        if hasattr(request, 'user') and request.user.is_authenticated:
            context = SimpleNamespace(user=request.user)
        else:
            context = SimpleNamespace(user=None)
        return context
```

File: govstack_api/building_blocks/bb_digital_registries/registries/base_registry.py (hunk 4 above)
Comment (lruzicki) [reply govstack_api/building_blocks/bb_digital_registries/registries/base_registry.py]: Ok, done

**Review by dborowiecki — APPROVED** [review]
